subsys/net/ip/tcp.c

```diff
@@ -361,8 +361,12 @@ static int tcp_conn_unref_debug(struct tcp *conn, const char *caller, int line)
 static int tcp_conn_unref(struct tcp *conn)
 #endif
 {
-	int ref_count = atomic_get(&conn->ref_count);
 	struct net_pkt *pkt;
+	int ref_count = atomic_dec(&conn->ref_count) - 1;
```

Comment:
And move the decrement and check just before the mutex lock, so that we don't check other logic.